File: app/src/main/java/algolia/com/demo/moviesearch/ui/ResultsListView.java
package algolia.com.demo.moviesearch.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.widget.ListView;

import com.algolia.instantsearch.model.AlgoliaResultsListener;
import com.algolia.instantsearch.model.SearchResults;

import java.util.List;

import algolia.com.demo.moviesearch.R;
import algolia.com.demo.moviesearch.io.SearchResultsJsonParser;
import algolia.com.demo.moviesearch.model.HighlightedResult;
import algolia.com.demo.moviesearch.model.Movie;

public class ResultsListView extends ListView implements AlgoliaResultsListener {
    private final MovieAdapter adapter;
    private SearchResultsJsonParser resultsParser = new SearchResultsJsonParser();

    public ResultsListView(final Context context, AttributeSet attrs) {
        super(context, attrs);
        setAdapter(adapter = new MovieAdapter(context, R.layout.cell_movie));
    }

    @Override public void onResults(@NonNull SearchResults results, boolean isLoadingMore) {
        if (!isLoadingMore) {
            List<HighlightedResult<Movie>> resultList = resultsParser.parseResults(results.content);
            adapter.clear();
            adapter.addAll(resultList);
            // Scroll the list back to the top.
            smoothScrollToPosition(0);
        } else {
            List<HighlightedResult<Movie>> resultList = resultsParser.parseResults(results.content);
            adapter.addAll(resultList);
        }
    }
}

File: app/src/main/java/algolia/com/demo/moviesearch/ui/ResultsListView.kt
package algolia.com.demo.moviesearch.ui

import algolia.com.demo.moviesearch.R
import algolia.com.demo.moviesearch.io.SearchResultsJsonParser
import android.content.Context
import android.util.AttributeSet
import android.widget.ListView
import com.algolia.instantsearch.model.AlgoliaErrorListener
import com.algolia.instantsearch.model.AlgoliaResultListener
import com.algolia.instantsearch.model.SearchResults
import com.algolia.search.saas.AlgoliaException
import com.algolia.search.saas.Query

class ResultsListView(context: Context, attrs: AttributeSet) : ListView(context, attrs),
        AlgoliaResultListener, AlgoliaErrorListener {
    private val adapter: MovieAdapter = MovieAdapter(context, R.layout.cell_movie)
    private val resultsParser = SearchResultsJsonParser()

    init {
        setAdapter(adapter)
    }

    override fun onResults(results: SearchResults, isLoadingMore: Boolean) {
        if (!isLoadingMore) {
            val resultList = resultsParser.parseResults(results.content)
            adapter.clear()
            adapter.addAll(resultList)
            // Scroll the list back to the top.
            smoothScrollToPosition(0)
        } else {
            val resultList = resultsParser.parseResults(results.content)
            adapter.addAll(resultList)
        }
    }

    override fun onError(query: Query, error: AlgoliaException) {}
}
